# Đồ án cuối kỳ môn Lập trình cho Khoa học dữ liệu
## Giảng viên hướng dẫn: 
### Thầy Trần Trung Kiên
### Cô Phan Thị Phương Uyên
## Sinh viên thực hiện:
##### Dương Hoàng Quang Khanh, mã sinh viên: 1712521

---

## Import các thư viện cần thiết

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

## Về nguồn dữ liệu:
Dữ liệu được sử dụng trong bài tập này là dữ liệu về 20 nhóm mặt hàng xuất khẩu và 20 nhóm mặt hàng xuất khẩu chủ lực của Việt Nam với các quốc gia trên thế giới trong giai đoạn 2010-2019. Dữ liệu được thống kê bởi Phòng ban thống kê Liên Hợp Quốc (United Nations Statistics Division) và được em trích xuất và tải về từ trang https://comtrade.un.org/data. Dữ liệu được thống kê theo từng năm và được lưu trong file "comtrade_data.csv". 

## Khám phá dữ liệu
### Đọc dữ liệu từ file

In [51]:
pd.set_option('display.max_columns', None) #Giúp hiển thị tất cả các cột của dữ liệu
data = pd.read_csv("comtrade_data.csv")
data.head(10)

Classification  Year  Period  Period Desc.  Aggregate Level  \
0             H3  2010    2010          2010                2   
1             H3  2010    2010          2010                2   
2             H3  2010    2010          2010                2   
3             H3  2010    2010          2010                2   
4             H3  2010    2010          2010                2   
5             H3  2010    2010          2010                2   
6             H3  2010    2010          2010                2   
7             H3  2010    2010          2010                2   
8             H3  2010    2010          2010                2   
9             H3  2010    2010          2010                2   

   Trade Flow Code Trade Flow  Reporter Code  Reporter Reporter ISO  \
0                0          M            704  Viet Nam          VNM   
1                0          M            704  Viet Nam          VNM   
2                0          M            704  Viet Nam          VNM   
3                0          M            704  Viet Nam          VNM   
4                0          M            704  Viet Nam          VNM   
5                0          M            704  Viet Nam          VNM   
6                0          M            704  Viet Nam          VNM   
7                0          M            704  Viet Nam          VNM   
8                0          M            704  Viet Nam          VNM   
9                0          M            704  Viet Nam          VNM   

   Partner Code    Partner Partner ISO  Commodity Code  \
0            32  Argentina         ARG               2   
1            36  Australia         AUS               2   
2            40    Austria         AUT               2   
3            76     Brazil         BRA               2   
4           124     Canada         CAN               2   
5           152      Chile         CHL               2   
6           156      China         CHN               2   
7           208    Denmark         DNK               2   
8           251     France         FRA               2   
9           348    Hungary         HUN               2   

                    Commodity  Qty Unit Code  Qty Unit  Qty  \
0  Meat and edible meat offal             -1       NaN  NaN   
1  Meat and edible meat offal             -1       NaN  NaN   
2  Meat and edible meat offal             -1       NaN  NaN   
3  Meat and edible meat offal             -1       NaN  NaN   
4  Meat and edible meat offal             -1       NaN  NaN   
5  Meat and edible meat offal             -1       NaN  NaN   
6  Meat and edible meat offal             -1       NaN  NaN   
7  Meat and edible meat offal             -1       NaN  NaN   
8  Meat and edible meat offal             -1       NaN  NaN   
9  Meat and edible meat offal             -1       NaN  NaN   

   Alt Qty Unit Code  Alt Qty Unit  Alt Qty  Netweight (kg)  \
0                 -1           NaN      NaN             NaN   
1                 -1           NaN      NaN             NaN   
2                 -1           NaN      NaN             NaN   
3                 -1           NaN      NaN             NaN   
4                 -1           NaN      NaN             NaN   
5                 -1           NaN      NaN             NaN   
6                 -1           NaN      NaN             NaN   
7                 -1           NaN      NaN             NaN   
8                 -1           NaN      NaN             NaN   
9                 -1           NaN      NaN             NaN   

   Gross weight (kg)  Trade Value (US$)  CIF Trade Value (US$)  \
0                NaN            1705860              1705860.0   
1                NaN           10492280             10492280.0   
2                NaN              99310                99310.0   
3                NaN            5473690              5473690.0   
4                NaN             552850               552850.0   
5                NaN               1420                 1420.0   
6              

### Kích thước của dữ liệu

In [9]:
print(data.shape)

(31520, 28)


Có thể thấy dữ liệu gồm có 31520 dòng và 28 cột.

Mỗi dòng dữ liệu mô tả về giá trị xuất khẩu hoặc nhập khẩu của Việt Nam với một đối tác về một nhóm mặt hàng trong một năm duy nhất.

Về mô tả thông tin của mỗi cột xin mời xem mô tả dưới đây:

- **Classification**: các phân loại quốc tế áp dụng cho dữ liệu thương mại và thuế quan được đưa ra bởi tổ chức Hải quan thế giới (WCO), H3 áp dụng cho hàng hóa từ 2007 - 2012, H4 là từ 2012 - 2017, và H5 là từ 2017 đến nay.
- **Year, Period, Period Desc**: Vì dữ liệu thông kê theo từng năm nên cả ba cột này đều mô tả năm mà dữu liệu được thống kê.
- **Aggregate Level**: đề cập đến mức độ chi tiết trong thống kê dữ liệu.
- **Trade Flow Code, Trade Flow**: phân biệt giữa mặt hàng xuất khẩu và nhập khẩu. 
- **Reporter Code, Reporter, Reporter ISO**: lần lượt là Mã số, Tên, Viết tắt của quốc gia được chọn là chủ thể thu thập thông tin. Ở đây là Việt Nam.
- **Partner Code, Partner, Partner ISO**: lần lượt là Mã số, Tên Viết tắt của các quốc gia đối tác.
- **Qty Unit Code, Qty Unit, Qty, Alt Qty Unit Code, Alt Qty Unit, Alt Qty, Netweight (kg), Gross weight (kg)**: Các cột này nói về các đơn vị số lượng và khối lượng hàng hóa, nhưng vì ở đây người ta chỉ thống kê về giá trị nên cá cột này không mang ý nghĩa.
- **Trade Value, CIF Trade Value, FOB Trade Value**: được tính theo US$, là tổng giá trị trong năm đó.

### Loại bỏ trước một số cột không mang lại giá trị thông tin

Các giá trị trong một số cột mà ta xem xét sẽ bỏ.

In [54]:
data["Classification"].unique()
c = ["Classification", "Reporter", "Reporter Code", "Reporter ISO", "Qty Unit Code",
     "Qty Unit", "Qty", "Alt Qty Unit Code", "Alt Qty Unit", "Alt Qty", "Netweight (kg)", "Gross weight (kg)"]
unq = pd.Series([data[i].unique() for i in c], index=c)
unq

Classification       [H3, H4, H5]
Reporter               [Viet Nam]
Reporter Code               [704]
Reporter ISO                [VNM]
Qty Unit Code                [-1]
Qty Unit                    [nan]
Qty                    [nan, 0.0]
Alt Qty Unit Code            [-1]
Alt Qty Unit                [nan]
Alt Qty                [nan, 0.0]
Netweight (kg)         [nan, 0.0]
Gross weight (kg)      [nan, 0.0]
dtype: object

- Cột **"Classification"** này chỉ có 3 giá trị là 'H3', 'H4', 'H5'. Nó biểu thị cho hàng hóa giao dịch trong giai đoạn nào ứng với 3 giá trị này. Vì trong dữ liệu này ta đã có thông tin về năm giao dịch ở cột **"Year"**, thậm chí cột này còn chi tiết về năm hơn, do đó ta có thể bỏ đi cột **"Classification"**.
- Ba cột **"Reporter", "Reporter ISO", "Partner Code"** đều chỉ chứa 1 giá trị trong mỗi cột (liên quan đến Việt Nam) vì trong dữ liệu này thì Việt Nam là quốc gia được thống kê, do đó ta có thể bỏ ba cột này.
- Các cột **Qty Unit Code, Qty Unit, Qty, Alt Qty Unit Code, Alt Qty Unit, Alt Qty, Netweight (kg), Gross weight (kg)** chỉ chứ các giá trị không có ý nghĩa nên ta cũng sẽ bỏ.
- Ba cột là **"Year", "Period", "Period Desc."** đều cùng mang một giá trị là năm thống kê nên ta chỉ cần giữ lại một cột là **"Year"** và bỏ đi hai cột **"Period", "Period Desc."**.
- Hai cột cùng biểu diễn các giá trị phân biệt Xuất khẩu với nhập khẩu là **"Trade Flow Code"**và **"Trade Flow"** chỉ cần giữ lại 1 cột.
- Các cột **"Partner Code", "Partner", "Partner ISO"** đều biểu diễn cùng một quốc gia đối tác trên một dòng, nên ta có thể bỏ đi hai cột là **"Partner Code", "Partner ISO"** và giữ lại cột **"Partner""**
- Một điều đáng lưu ý Mới các mặt hàng nhập khẩu thì **"Trade value"** sẽ bằng **"CIF Trade Value"**, ngược lại với hàng hóa xuất khẩu thì **"Trade Value"** sẽ bằng **"FOB Trade Value"**. Dó đó ta chỉ cẩn giữ lại cột **Trade Value**.

Các cột ta sẽ bỏ đi đó là "Classification", "Reporter", "Reporter Code", "Reporter ISO", "Qty Unit Code", "Qty Unit", "Qty", "Alt Qty Unit Code", "Alt Qty Unit", "Alt Qty", "Netweight (kg)", "Gross weight (kg)","Period", "Period Desc.", "Partner Code", "Partner ISO", "CIF Trade Value", "FOB Trade Value".

In [58]:
drop_cols = ["Classification", "Trade Flow Code","Reporter", "Reporter Code", "Reporter ISO", "Qty Unit Code",
             "Qty Unit", "Qty", "Alt Qty Unit Code", "Alt Qty Unit", "Alt Qty", "Netweight (kg)", "Gross weight (kg)", 
             "Period", "Period Desc.", "Partner Code", "Partner ISO", "CIF Trade Value (US$)", "FOB Trade Value (US$)"]
data.drop(labels=drop_cols, axis=1, inplace=True)
data.head(5)

Year  Aggregate Level Trade Flow    Partner  Commodity Code  \
0  2010                2          M  Argentina               2   
1  2010                2          M  Australia               2   
2  2010                2          M    Austria               2   
3  2010                2          M     Brazil               2   
4  2010                2          M     Canada               2   

                    Commodity  Trade Value (US$)  Flag  
0  Meat and edible meat offal            1705860     0  
1  Meat and edible meat offal           10492280     0  
2  Meat and edible meat offal              99310     0  
3  Meat and edible meat offal            5473690     0  
4  Meat and edible meat offal             552850     0